Throw linting errors if using non-polyfilled APIs

MrTimscampi opened the pull request:
**Changes**

This adds an ESLint plugin that provides rules to detect whenever a non-polyfilled API is used and throws an error.

It uses the Browserslist targets to detect the minimal versions of the browsers, then uses caniuse to lookup feature availability.

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

dkanada commented on src/components/input/gamepadtokey.js:
```diff
@@ -401,7 +403,7 @@ require(['apphost'], function (appHost) {
         // We want the gamepad to provide gamepad VK keyboard events rather than moving a
         // mouse like cursor. Set to "keyboard", the gamepad will provide such keyboard events
-        // and provide input to the DOM navigator.getGamepads API.
+        // and provide input to tonFocusOrGamepadDetachhe DOM navigator.getGamepads API.
```
```suggestion
        // and provide input to onFocusOrGamepadDetachhe DOM navigator.getGamepads API.
```

dmitrylyzo commented on src/components/input/keyboardnavigation.js:
```diff
@@ -159,7 +159,9 @@ function attachGamepadScript(e) {
 
 // No need to check for gamepads manually at load time, the eventhandler will be fired for that
-window.addEventListener("gamepadconnected", attachGamepadScript);
+if (navigator.getGamepads) { /* eslint-disable-line compat/compat */
```
`attachGamepadScript` above can also be placed inside.

MrTimscampi replied:
ESLint yells at me if I move the function inside (https://eslint.org/docs/rules/no-inner-declarations).

dmitrylyzo replied:
Oh, right. `var foo = function (...)` should work. Or we could insert it directly as `addEventListener` callback.

MrTimscampi replied:
I don't think it's really necessary to get it inside of there, it'll never get executed anyway if there is no gamepad support.

If you really feel like we should, would something like this work?

```javascript
window.addEventListener("gamepadconnected", function () {
    console.log("Gamepad connected! Attaching gamepadtokey.js script");
    window.removeEventListener("gamepadconnected", this);
    require(["components/input/gamepadtokey"]);
});
```

dmitrylyzo replied:
> would something like this work?

Nope. `this` is window. :disappointed: I didn't thought about `removeEventListener`.

OK. Leave it as it is.

MrTimscampi commented on src/components/appRouter.js:
```diff
@@ -268,6 +268,7 @@ define(['loading', 'globalize', 'events', 'viewManager', 'layoutManager', 'skinM
     }
 
     function getMaxBandwidth() {
+        /* eslint-disable compat/compat */
```
ESLint throws an error on `navigator.connection`, which is an API that allows to detect connection information (speed, type of conneciton, roaming status, etc).

I haven't found a polyfill and the feature is checked anyway, so I'm disabling the compat check for this block.

MrTimscampi commented on src/components/apphost.js:
```diff
@@ -453,6 +453,7 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
 
     if (doc) {
         doc.addEventListener(visibilityChange, function () {
+            /* eslint-disable-next-line compat/compat */
```
ESLint throws an error on Document.visibilityState, which I didn't find a polyfill for.

As the feature is conditional anyway, I disabled the warning.

JustAMan replied:
but the condition is on `doc`, not `doc.visibilityState`?

MrTimscampi replied:
`document[visibilityState]` resolves to `document.visibilityState`.

I changed it to make it clearer.

MrTimscampi commented on src/components/htmlMediaHelper.js:
```diff
@@ -31,7 +31,7 @@ define(['appSettings', 'browser', 'events'], function (appSettings, browser, eve
 
     function enableHlsShakaPlayer(item, mediaSource, mediaType) {
-
+        /* eslint-disable-next-line compat/compat */
```
window.MediaSource is not available on Chrome 27. As the feature is conditional anyway, I disabled the compat check to prevent ESLint from throwing an error.

MrTimscampi commented on src/components/input/gamepadtokey.js:
```diff
@@ -184,7 +184,7 @@ require(['apphost'], function (appHost) {
 
         // This would be nice but always seems to return true with electron
-        if (!isElectron && document.hidden) {
+        if (!isElectron && document.hidden) { /* eslint-disable-line compat/compat */
```
The Page Visibility API (document.hidden) isn't available on Chrome 27. As it's a conditional feature, I disabled the ESLint check.

MrTimscampi commented on src/components/input/gamepadtokey.js:
```diff
@@ -373,13 +373,15 @@ require(['apphost'], function (appHost) {
     }
 
     function onFocusOrGamepadAttach(e) {
+        /* eslint-disable-next-line compat/compat */
```
document.hasFocus() is throwing an error for Chrome 27.

It is supported [according to CanIUse](https://caniuse.com/#feat=mdn-api_document_hasfocus).

I opened an issue here: https://github.com/amilajack/eslint-plugin-compat/issues/304

MrTimscampi commented on src/scripts/browser.js:
```diff
@@ -292,6 +292,7 @@ define([], function () {
     }
 
     if (typeof document !== 'undefined') {
+        /* eslint-disable-next-line compat/compat */
```
navigator.maxTouchPoints isn't available on Chrome 27.

It's ignored because the feature is conditional.

MrTimscampi commented on src/components/input/gamepadtokey.js:
```diff
@@ -254,7 +254,7 @@ require(['apphost'], function (appHost) {
     function runInputLoop() {
         // Get the latest gamepad state.
-        var gamepads = navigator.getGamepads();
+        var gamepads = navigator.getGamepads(); /* eslint-disable-line compat/compat */
```
navigator.getGamepads isn't polyfill-able. This file is loaded conditionally anyway, now, do no issue in disabling the compat lint here.

MrTimscampi commented on src/controllers/playback/videoosd.js:
```diff
@@ -1152,7 +1152,7 @@ define(["playbackManager", "dom", "inputManager", "datetime", "itemHelper", "med
                 case "GamepadLeftThumbstickLeft":
                     // Ignores gamepad events that are always triggered, even when not focused.
-                    if (document.hasFocus()) {
+                    if (document.hasFocus()) { /* eslint-disable-line compat/compat */
```
Same linting issue as the other document.hasFocus(). An issue has been filed upstream to the plugin's repo.

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -453,7 +453,8 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
     if (doc) {
         doc.addEventListener(visibilityChange, function () {
-            if (document[visibilityState]) {
+            /* eslint-disable-next-line compat/compat */
+            if (document.visibilityState) {
```
`document[visibilityState]` was use `visibilityState` variable that set above. `document.visibilityState` is supposed be polyfilled.

MrTimscampi replied:
Hm, this entire thing seems a little screwy and convoluted.

It should work in a different way, probably. I'll try refactoring it all tonight.

MrTimscampi replied:
Should be better now :)

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,35 +426,27 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
-            if (document.visibilityState) {
+    if (self.document) {
+        document.addEventListener(visibilityChange, function () {
+            if (visibilityState) {
```
If `visibilityState` is set, `onAppHidden` is called in all cases. I tested by switching tabs in browser.
```suggestion
            if (document[visibilityState]) {
```
But it is strange: are all vendor-members boolean (as `document.hidden`)?

dmitrylyzo replied:
It seems that `mozHidden`, `msHidden` and `webkitHidden` must be used.
https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API in **Example**

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,33 +426,26 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
-    if (doc) {
-        if (void 0 !== doc.visibilityState) {
+    if (self.document) {
+        if (self.document.visibilityState !== undefined) {
```
```suggestion
        if (self.document.hidden !== undefined) {
```

MrTimscampi replied:
We're checking for API support here, not status, I believe?

MrTimscampi replied:
Alright, I stumbled on the source for that part of the code on MDN, it's a copy-paste from the "Example" section here: https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API

dmitrylyzo replied:
I think we are just choosing proper field to check visibility state.

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,33 +426,26 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
-            }
+        } else if (self.document.mozHidden !== undefined) {
+            visibilityChange = "mozvisibilitychange";
+            visibilityState = "mozVisibilityState";
```
```suggestion
            visibilityState = "mozHidden";
```

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,33 +426,26 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
+            visibilityState = "mozVisibilityState";
+        } else if (self.document.msHidden !== undefined) {
+            visibilityChange = "msvisibilitychange";
+            visibilityState = "msVisibilityState";
```
```suggestion
            visibilityState = "msHidden";
```

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,33 +426,26 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
+            visibilityState = "msVisibilityState";
+        } else if (self.document.webkitHidden !== undefined) {
+            visibilityChange = "webkitvisibilitychange";
+            visibilityState = "webkitVisibilityState";
```
```suggestion
            visibilityState = "webkitHidden";
```

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,33 +426,26 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
-        doc.addEventListener(visibilityChange, function () {
+    if (self.document) {
+        document.addEventListener(visibilityChange, function () {
             if (document[visibilityState]) {
```
ESLint considers `document.visibilityState` to be requested. :confused: 
```suggestion
            /* eslint-disable-next-line compat/compat */
            if (document[visibilityState]) {
```

MrTimscampi commented:
@dmitrylyzo I switched the whole thing to the version on MDN, since that's likely more correct than whatever was in there :)

I also removed the MS-specific case, since Edge 12+ supports it and that version is now ancient.

dmitrylyzo commented on src/components/apphost.js:
```diff
@@ -426,39 +424,27 @@ define(["appSettings", "browser", "events", "htmlMediaHelper", "webSettings"], f
-    if (doc) {
-        doc.addEventListener(visibilityChange, function () {
-            if (document[visibilityState]) {
+    if (self.document) {
```
I wonder why `self.document` is even checked - `document` is used everywhere.

MrTimscampi replied:
Good point. I removed it since it shouldn't matter anyway.
